docs/library/esp32.rst

```diff
@@ -153,6 +153,133 @@ Constants
+      - ``threshold1``: sets the counter value for the
+        ``PCNT.IRQ_THRESHOLD1`` event (see ``irq`` method)
+      - ``reset``: if ``True``, reset the counter value
+
+    The hardware initialisation is done in stages and so some of the keyword
+    arguments can be used in groups or in isolation to partially reconfigure a
+    unit:
+
+      - the ``pin`` keyword (optionally combined with ``mode_pin``) can be used
+        to change just the bound pin(s)
+      - ``rising`, ``falling``, ``mode_low`` and ``mode_high`` can be used
+        (singly or together) to change the counting logic - omitted keywords
+        use their default (``PCNT.IGNORE`` or ``PCNT.NORMAL``)
+      - ``filter`` can be used to change only the pulse width filter (with 0
+        disabling it)
+      - each of ``minimum``, ``maximum``, ``threshold0`` and ``threshold1`` can
+        be used to change these limit/event values individually; however,
+        setting any will reset the counter to zero (i.e., they imply
+        ``value=0``)
+
+    Each pulse counter unit supports two channels, 0 and 1, each able to
+    monitor different pins with different counting logic but updating the same
+    counter value. Use ``channel=1`` with the ``pin``, ``rising``, ``falling``,
+    ``mode_pin``, ``mode_low`` and ``mode_high`` keywords to configure the
+    second channel.
+
+    The second channel can be used to configure 4X quadrature decoding with a
+    single counter unit::
+
+        pin_a = Pin(2, Pin.INPUT, pull=Pin.PULL_UP)
+        pin_b = Pin(3, Pin.INPUT, pull=Pin.PULL_UP)
+        rotary = PCNT(0, minimum=-32000, maximum=32000)
+        rotary.init(channel=0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE)
+        rotary.init(channel=1, pin=pin_b, falling=PCNT.DECREMENT, rising=PCNT.INCREMENT, mode_pin=pin_a, mode_low=PCNT.REVERSE)
+        rotary.start()
+
+.. method:: PCNT.value([reset])
+
+    Call this method with no arguments to return the current counter value.
+
+    If the optional *reset* argument is set to ``True`` then the counter is
+    reset (but the previous value is returned). Read and reset is not atomic and
+    so it is possible for a pulse to be missed.
+
+.. method:: PCNT.irq(handler=None, trigger=PCNT.IRQ_ZERO)
+
+    ESP32 pulse counters support interrupts on these counter events:
+
+      - ``PCNT.IRQ_ZERO``: the counter has reset to zero
+      - ``PCNT.IRQ_MINIMUM``: the counter has hit the ``minimum`` value
+      - ``PCNT.IRQ_MAXIMUM``: the counter has hit the ``maximum`` value
+      - ``PCNT.IRQ_THRESHOLD0``: the counter has hit the ``threshold0`` value
+      - ``PCNT.IRQ_THRESHOLD1``: the counter has hit the ``threshold1`` value
+
+    ``trigger`` should be the desired events or'ed together. The ``handler``
+    function should take a single argument which will be a bit mask indicating
+    which counter event(s) occurred.
```

Comment:
This is not how other machine irq methods work.  The "standard" (see eg rp2's Pin.irq) is for the irq handler to take a single argument being the instance of the original peripheral object, in this case the PCNT instance.

Also, the `irq()` method should return a special IRQ object which can be used to get the flags indicating which events occurred.  This is all implemented by the common `shared/runtime/mpirq.c` code (which isn't as widely used as it should be...).

Reply:
I didn't even realise that callback objects supported reading the triggered event! This doesn't seem to be well documented (the **esp32** `machine.Pin.irq()` docs say nothing more than "This method returns a callback object"), and I hadn't realised there was common code for supporting it.

Reply:
Using `mpirq.c` also means we can no longer use the static scheduler node.

It's not super obvious to me how to add support for using static scheduler nodes to `mpirq.c` without having a node-per-instance (i.e. one per PCNT).

Reply:
> Using mpirq.c also means we can no longer use the static scheduler node.

Bah! All bets are off then 😉

Reply:
> Perhaps just zero the flags after the handler has been called instead?

That was my first thought, but unfortunately this isn't possible with `mpirq.c` because it calls the handler for you. What you'd need to do is to have a wrapper handler (owned by esp32_pcnt_obj_t) which the handler (owned by esp32_pcnt_irq_obj_t) called (which then reset the flags).

But this still doesn't work because the low-level ISR might update the flags in the meantime.

This might be a good opportunity to re-evaluate how `mpirq.c` works -- in other places where we use mpirq.c I think it would benefit from changing to this behavior. e.g. in `rp2/machine_pin.c` the `gpio_irq` function always sets the flags, so for example a `RISING` will clobber a pending `FALLING`.

> I was pretty sure at the time that a call to .value() within the request handler wouldn't cause a loop in my original code, but I'd have to diff your PR against my original branch and then stare at it a bit to see if that still holds.

I did see https://github.com/micropython/micropython/pull/7582#issuecomment-1165457273 but I think https://github.com/micropython/micropython/pull/7582/files#r1015829258 is more relevant (and I think here you're saying that it can cause a loop?). If the ISR runs (and sets more flags) then it will keep recursing. Like you say, this isn't a problem if the irq handler method is re-entrant (which it should be?), and it's also exceptionally unlikely to happen in practice (also now it's confined to a single unit).

> Bah! All bets are off then 😉

Indeed. Especially if you have multiple PCNT units running concurrently. I'll leave this one for @dpgeorge to comment before I start modifying `mpirq.c` to support static nodes. (I really don't want to duplicate the functionality `mpirq.c` in `esp32_pcnt.c` just to have a single shared static node).

Reply:
> I did see https://github.com/micropython/micropython/pull/7582#issuecomment-1165457273 but I think https://github.com/micropython/micropython/pull/7582/files#r1015829258 is more relevant (and I think here you're saying that it can cause a loop?)

Ah, yes. What I figured was that if a second event occurred immediately after starting the call to the handler, but before a call to `.value()` within the handler, that *would* cause a further nested call to the handler, but it could only recurse further if interrupts kept happening.

So an interrupt storm could conceivably cause you to run out of stack, but it couldn't get into an infinite loop. Of course, nested calls to the handler function might cause some other ugly behaviour in the user code…

I was planning to put more thought into how to avoid this, but my brain started to hurt.

Maybe the real problem here is the large-number implementation of `Counter()`, which requires an offset that is kept in careful sync with the hardware counter. This behaviour could, perhaps, be pushed back into `PCNT()` itself where more careful interrupt handling could be done in C.